### ISSUE TITLE
d1: Scaling Reasoning in Diffusion Large Language Models via
  Reinforcement Learning, Siyan Zhao+, arXiv'25

### ISSUE THREAD

**Issue by AkihikoWatanabe** [AkihikoWatanabe/paper_notes#1893]
# URL
- https://arxiv.org/abs/2504.12216
# Authors
  - Siyan Zhao
  - Devaansh Gupta
  - Qinqing Zheng
  - Aditya Grover
# Abstract
  - Recent large language models (LLMs) have demonstrated strong reasoning capabilities that benefits from online reinforcement learning (RL). These capabilities have primarily been demonstrated within the left-to-right autoregressive (AR) generation paradigm. In contrast, non-autoregressive paradigms based on diffusion generate text in a coarse-to-fine manner. Although recent diffusion-based large language models (dLLMs) have achieved competitive language modeling performance compared to their AR counterparts, it remains unclear if dLLMs can also leverage recent advances in LLM reasoning. To this end, we propose d1, a framework to adapt pre-trained masked dLLMs into reasoning models via a combination of supervised finetuning (SFT) and RL. Specifically, we develop and extend techniques to improve reasoning in pretrained dLLMs: (a) we utilize a masked SFT technique to distill knowledge and instill self-improvement behavior directly from existing datasets, and (b) we introduce a novel critic-free, policy-gradient based RL algorithm called diffu-GRPO. Through empirical studies, we investigate the performance of different post-training recipes on multiple mathematical and logical reasoning benchmarks. We find that d1 yields the best performance and significantly improves performance of a state-of-the-art dLLM.
# Translation (by gpt-4o-mini)
- 最近の大規模言語モデル（LLMs）は、オンライン強化学習（RL）から恩恵を受けた強力な推論能力を示しています。これらの能力は主に、左から右への自己回帰（AR）生成パラダイム内で実証されています。それに対して、拡散に基づく非自己回帰パラダイムは、粗から細への方法でテキストを生成します。最近の拡散ベースの大規模言語モデル（dLLMs）は、ARモデルと比較して競争力のある言語モデリング性能を達成していますが、dLLMsがLLM推論の最近の進展を活用できるかどうかは不明です。この目的のために、我々はd1を提案します。これは、事前学習されたマスク付きdLLMsを、教師ありファインチューニング（SFT）とRLの組み合わせを通じて推論モデルに適応させるフレームワークです。具体的には、事前学習されたdLLMsの推論を改善するための技術を開発・拡張します：(a) 既存のデータセットから知識を抽出し、自己改善行動を直接植え付けるためにマスク付きSFT技術を利用し、(b) diffu-GRPOと呼ばれる新しいクリティックフリーのポリシー勾配ベースのRLアルゴリズムを導入します。実証研究を通じて、複数の数学的および論理的推論ベンチマークにおける異なるポストトレーニングレシピの性能を調査します。その結果、d1が最良の性能を発揮し、最先端のdLLMの性能を大幅に向上させることがわかりました。
# Summary (by gpt-4o-mini)
- d1というフレームワークを提案し、マスク付きdLLMsを教師ありファインチューニングと強化学習で推論モデルに適応。マスク付きSFT技術で知識を抽出し、diffu-GRPOという新しいRLアルゴリズムを導入。実証研究により、d1が最先端のdLLMの性能を大幅に向上させることを確認。

**Comment by AkihikoWatanabe** [comment]
元ポスト:https://x.com/iscienceluvr/status/1912785180504535121?s=46&t=Y6UuIHB0Lv0IpmFAjlc2-Q

**Comment by AkihikoWatanabe** [comment]
dLLMに対してGRPOを適用する手法(diffuGRPO)を提案している。
long CoTデータでSFTしてreasoning capabilityを強化した後、diffuGRPOで追加のpost-trainingをしてさらに性能をboostする。

**Comment by AkihikoWatanabe** [comment]
GRPOではtoken levelの尤度とsequence全体の尤度を計算する必要があるが、dLLMだとautoregressive modelのようにchain ruleを適用する計算方法はできないので、効率的に尤度を推定するestimatorを用いてGPPOを適用するdiffuGRPOを提案している。

diffuGRPO単体でも、8BモデルだがSFTよりも性能向上に成功している。SFTの後にdiffuGRPOを適用するとさらに性能が向上する。

SFTではs1 #1749 で用いられたlong CoTデータを用いている。しっかり理解できていないが、diffuGRPO+verified rewardによって、long CoTの学習データを用いなくても、安定してreasoning能力を発揮することができようになった、ということなのだろうか？
しかし、AppendixCを見ると、元々のLLaDAの時点でreasoning traceを十分な長さで出力しているように見える。もしLLaDAが元々long CoTを発揮できたのだとしたら、long CoTできるようになったのはdiffuGRPOだけの恩恵ではないということになりそうだが、LLaDAは元々long CoTを生成できるようなモデルだったんだっけ…？その辺追えてない（dLLMがメジャーになったら追う）。
